src/language-js/printer-estree.js

```diff
@@ -4873,8 +4871,21 @@ function printMemberChain(path, options, print) {
   // If the last group is a function it's okay to inline if it fits.
   if (
     hasComment ||
-    callExpressionCount >= 3 ||
-    printedGroups.slice(0, -1).some(willBreak)
+    callExpressions.length >= 3 ||
+    printedGroups.slice(0, -1).some(willBreak) ||
+    /**
+     *     scopes.filter(scope => scope.value !== '').map((scope, i) => {
+     *       // multi line content
+     *     })
+     */
+    (((lastGroupDoc, lastGroupNode) =>
+      isCallOrOptionalCallExpression(lastGroupNode) && willBreak(lastGroupDoc))(
+      getLast(printedGroups),
+      getLast(getLast(groups)).node
+    ) &&
+      callExpressions
+        .slice(0, -1)
+        .some(n => n.arguments.some(isFunctionOrArrowExpression)))
```

Comment:
I don't understand why this is necessary. Did the changes below break this case (`foo.call().call(x => { /* multi line */ }`) ?

Reply:
```diff
diff --git a/src/language-js/printer-estree.js b/src/language-js/printer-estree.js
index 1d3e0040..0a142484 100644
--- a/src/language-js/printer-estree.js
+++ b/src/language-js/printer-estree.js
@@ -4882,10 +4882,12 @@ function printMemberChain(path, options, print) {
       isCallOrOptionalCallExpression(lastGroupNode) && willBreak(lastGroupDoc))(
       getLast(printedGroups),
       getLast(getLast(groups)).node
-    ) &&
-      callExpressions
-        .slice(0, -1)
-        .some(n => n.arguments.some(isFunctionOrArrowExpression)))
+    ) 
+    // &&
+    //   callExpressions
+    //     .slice(0, -1)
+    //     .some(n => n.arguments.some(isFunctionOrArrowExpression))
+        )
   ) {
     return group(expanded);
   }
diff --git a/tests/flow/async_iteration/__snapshots__/jsfmt.spec.js.snap b/tests/flow/async_iteration/__snapshots__/jsfmt.spec.js.snap
index ff0ec221..232ecc3a 100644
--- a/tests/flow/async_iteration/__snapshots__/jsfmt.spec.js.snap
+++ b/tests/flow/async_iteration/__snapshots__/jsfmt.spec.js.snap
@@ -140,9 +140,11 @@ declare var gen: AsyncGenerator<void, string, void>;
 
 // You can pass whatever you like to return, it doesn't need to be related to
 // the AsyncGenerator's return type
-gen.return(0).then(result => {
-  (result.value: void); // error: string | number ~> void
-});
+gen
+  .return(0)
+  .then(result => {
+    (result.value: void); // error: string | number ~> void
+  });
 
 // However, a generator can "refuse" the return by catching an exception and
 // yielding or returning internally.
diff --git a/tests/function_first_param/__snapshots__/jsfmt.spec.js.snap b/tests/function_first_param/__snapshots__/jsfmt.spec.js.snap
index 4a20306a..0f0a06b1 100644
--- a/tests/function_first_param/__snapshots__/jsfmt.spec.js.snap
+++ b/tests/function_first_param/__snapshots__/jsfmt.spec.js.snap
@@ -27,18 +27,20 @@ db.collection('indexOptionDefault').createIndex({ a: 1 }, {
   done();
 });~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 //https://github.com/prettier/prettier/issues/3002
-beep.boop().baz(
-  "foo",
-  {
-    some: {
-      thing: {
-        nested: true
+beep
+  .boop()
+  .baz(
+    "foo",
+    {
+      some: {
+        thing: {
+          nested: true
+        }
       }
-    }
-  },
-  { another: { thing: true } },
-  () => {}
-);
+    },
+    { another: { thing: true } },
+    () => {}
+  );
 
 //https://github.com/prettier/prettier/issues/2984
 db.collection("indexOptionDefault").createIndex(
diff --git a/tests/markdown/__snapshots__/jsfmt.spec.js.snap b/tests/markdown/__snapshots__/jsfmt.spec.js.snap
index 53bf6028..94a36643 100644
--- a/tests/markdown/__snapshots__/jsfmt.spec.js.snap
+++ b/tests/markdown/__snapshots__/jsfmt.spec.js.snap
@@ -1472,9 +1472,11 @@ If \`options.useCache\` is \`false\`, all caching will be bypassed.
 
 \`\`\`js
 const text = fs.readFileSync(filePath, "utf8");
-prettier.resolveConfig(filePath).then(options => {
-  const formatted = prettier.format(text, options);
-});
+prettier
+  .resolveConfig(filePath)
+  .then(options => {
+    const formatted = prettier.format(text, options);
+  });
 \`\`\`
 
 Use \`prettier.resolveConfig.sync([filePath [, options]])\` if you'd like to use
@@ -3443,9 +3445,11 @@ If \`options.useCache\` is \`false\`, all caching will be bypassed.
 
 \`\`\`js
 const text = fs.readFileSync(filePath, 'utf8');
-prettier.resolveConfig(filePath).then(options => {
-  const formatted = prettier.format(text, options);
-});
+prettier
+  .resolveConfig(filePath)
+  .then(options => {
+    const formatted = prettier.format(text, options);
+  });
 \`\`\`
 
 Use \`prettier.resolveConfig.sync([filePath [, options]])\` if you'd like to use
diff --git a/tests/method-chain/__snapshots__/jsfmt.spec.js.snap b/tests/method-chain/__snapshots__/jsfmt.spec.js.snap
index 5110b148..e366a31d 100644
--- a/tests/method-chain/__snapshots__/jsfmt.spec.js.snap
+++ b/tests/method-chain/__snapshots__/jsfmt.spec.js.snap
@@ -563,20 +563,22 @@ function f() {
 }
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 export default function theFunction(action$, store) {
-  return action$.ofType(THE_ACTION).switchMap(action =>
-    Observable.webSocket({
-      url: THE_URL,
-      more: stuff(),
-      evenMore: stuff({
-        value1: true,
-        value2: false,
-        value3: false
+  return action$
+    .ofType(THE_ACTION)
+    .switchMap(action =>
+      Observable.webSocket({
+        url: THE_URL,
+        more: stuff(),
+        evenMore: stuff({
+          value1: true,
+          value2: false,
+          value3: false
+        })
       })
-    })
-      .filter(data => theFilter(data))
-      .map(({ theType, ...data }) => theMap(theType, data))
-      .retryWhen(errors => errors)
-  );
+        .filter(data => theFilter(data))
+        .map(({ theType, ...data }) => theMap(theType, data))
+        .retryWhen(errors => errors)
+    );
 }
 
 function f() {
diff --git a/tests/performance/__snapshots__/jsfmt.spec.js.snap b/tests/performance/__snapshots__/jsfmt.spec.js.snap
index 7a601d13..47de7bf6 100644
--- a/tests/performance/__snapshots__/jsfmt.spec.js.snap
+++ b/tests/performance/__snapshots__/jsfmt.spec.js.snap
@@ -31,35 +31,63 @@ someObject.someFunction().then(function() {
   });
 });
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-someObject.someFunction().then(function() {
-  return someObject.someFunction().then(function() {
-    return someObject.someFunction().then(function() {
-      return someObject.someFunction().then(function() {
-        return someObject.someFunction().then(function() {
-          return someObject.someFunction().then(function() {
-            return someObject.someFunction().then(function() {
-              return someObject.someFunction().then(function() {
-                return someObject.someFunction().then(function() {
-                  return someObject.someFunction().then(function() {
-                    return someObject.someFunction().then(function() {
-                      return someObject.someFunction().then(function() {
-                        return someObject.someFunction().then(function() {
-                          return someObject.someFunction().then(function() {
-                            anotherFunction();
+someObject
+  .someFunction()
+  .then(function() {
+    return someObject
+      .someFunction()
+      .then(function() {
+        return someObject
+          .someFunction()
+          .then(function() {
+            return someObject
+              .someFunction()
+              .then(function() {
+                return someObject
+                  .someFunction()
+                  .then(function() {
+                    return someObject
+                      .someFunction()
+                      .then(function() {
+                        return someObject
+                          .someFunction()
+                          .then(function() {
+                            return someObject
+                              .someFunction()
+                              .then(function() {
+                                return someObject
+                                  .someFunction()
+                                  .then(function() {
+                                    return someObject
+                                      .someFunction()
+                                      .then(function() {
+                                        return someObject
+                                          .someFunction()
+                                          .then(function() {
+                                            return someObject
+                                              .someFunction()
+                                              .then(function() {
+                                                return someObject
+                                                  .someFunction()
+                                                  .then(function() {
+                                                    return someObject
+                                                      .someFunction()
+                                                      .then(function() {
+                                                        anotherFunction();
+                                                      });
+                                                  });
+                                              });
+                                          });
+                                      });
+                                  });
+                              });
                           });
-                        });
                       });
-                    });
                   });
-                });
               });
-            });
           });
-        });
       });
-    });
   });
-});
 
 `;

```

Reply:
🙈